src/derivative_operators/derivative_stencil.jl

```diff
@@ -0,0 +1,205 @@
+abstract type AbstractStencilOperator{T} <: AbstractDiffEqLinearOperator{T} end
+
+##################################################
+# Derivative stencil operator for uniform grid
+struct UniformDerivativeStencil{T,S<:SVector} <: AbstractStencilOperator{T}
+    derivative_order    :: Int
+    approximation_order :: Int
+    dimension           :: Tuple{Int,Int}
+    stencil_length      :: Int
+    stencil_coefs       :: S
+    function UniformDerivativeStencil(dorder,aorder,dx::T,dim_extended) where {T}
+        stencil_length = dorder + aorder - 1 + (dorder + aorder) % 2
+        stl_2 = div(stencil_length, 2)
+        dim = (dim_extended - stencil_length + 1, dim_extended)
+        stencil_coefs = convert(SVector{stencil_length,T}, calculate_weights(
+            dorder, zero(T), dx .* collect(-stl_2 : 1 : stl_2)))
+        new{T,typeof(stencil_coefs)}(dorder,aorder,dim,stencil_length,stencil_coefs)
+    end
+    UniformDerivativeStencil(xgrid::AbstractRange{T},dorder,aorder) where {T} =
+        UniformDerivativeStencil(dorder,aorder,step(xgrid),length(xgrid))
+end
+function mul!(y::AbstractVector{T}, L::UniformDerivativeStencil{T,S}, x::AbstractVector{T}) where {T,S}
+    coeffs = L.stencil_coefs
+    Threads.@threads for i in 1:length(y)
+        ytemp = zero(T)
+        @inbounds for idx in 1:L.stencil_length
+            ytemp += coeffs[idx] * x[i + idx - 1]
+        end
+        y[i] = ytemp
+    end
+    return y
+end
+function convert(::Type{AbstractMatrix}, L::UniformDerivativeStencil)
+    coeffs = L.stencil_coefs
+    mat = spzeros(eltype(L), size(L,1), size(L,2))
+    for i in 1:size(L,1)
+        for idx in 1:L.stencil_length
+            mat[i, i + idx - 1] = coeffs[idx]
```

Comment:
this is a very slow way of doing this, but we can ignore that for now. It would be better to try and make a BandedMatrix anyways.